spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamBinder.java

```diff
@@ -0,0 +1,188 @@
+			ExtendedProducerProperties<KStreamProducerProperties> properties) {
+		ExtendedProducerProperties<KafkaProducerProperties> extendedProducerProperties = new ExtendedProducerProperties<KafkaProducerProperties>(
+				new KafkaProducerProperties());
+		this.kafkaTopicProvisioner.provisionProducerDestination(name , extendedProducerProperties);
+		if (HeaderMode.embeddedHeaders.equals(properties.getHeaderMode())) {
+			outboundBindTarget = outboundBindTarget.map(new KeyValueMapper<Object, Object, KeyValue<Object, Object>>() {
+				@Override
+				public KeyValue<Object, Object> apply(Object k, Object v) {
+					if (v instanceof Message) {
+						try {
+							return new KeyValue<>(k, (Object)KStreamBinder.this.serializeAndEmbedHeadersIfApplicable((Message<?>) v));
+						}
+						catch (Exception e) {
+							throw new IllegalArgumentException(e);
+						}
+					}
+					else {
+						throw new IllegalArgumentException("Wrong type of message " + v);
+					}
+				}
+			});
+		}
+		else {
+			if (!properties.isUseNativeEncoding()) {
+				outboundBindTarget = outboundBindTarget
+						.map(new KeyValueMapper<Object, Object, KeyValue<Object, Object>>() {
+							@Override
+							public KeyValue<Object, Object> apply(Object k, Object v) {
+								return KeyValue.pair(k, (Object)KStreamBinder.this.serializePayloadIfNecessary((Message<?>) v));
+							}
+						});
+			}
+			else {
+				outboundBindTarget = outboundBindTarget
+						.map(new KeyValueMapper<Object, Object, KeyValue<Object, Object>>() {
+							@Override
+							public KeyValue<Object, Object> apply(Object k, Object v) {
+								return KeyValue.pair(k, ((Message<?>) v).getPayload());
+							}
+						});
+			}
+		}
+		if (!properties.isUseNativeEncoding() || StringUtils.hasText(properties.getExtension().getKeySerde()) || StringUtils.hasText(properties.getExtension().getValueSerde())) {
+			Serde<?> keySerde = Serdes.ByteArray();
+			Serde<?> valueSerde = Serdes.ByteArray();
+			try {
+				if (StringUtils.hasText(properties.getExtension().getKeySerde())) {
+					keySerde = Utils.newInstance(properties.getExtension().getKeySerde(), Serde.class);
+					if (keySerde instanceof Configurable) {
+						((Configurable) keySerde).configure(streamsConfig.originals());
+					}
+				}
+			}
+			catch (ClassNotFoundException e) {
+				throw new IllegalStateException("Serde class not found: ", e);
+			}
+			try {
+				if (StringUtils.hasText(properties.getExtension().getValueSerde())) {
+					valueSerde = Utils.newInstance(properties.getExtension().getValueSerde(), Serde.class);
+					if (valueSerde instanceof Configurable) {
+						((Configurable) valueSerde).configure(streamsConfig.originals());
+					}
+				}
+			}
+			catch (ClassNotFoundException e) {
+				throw new IllegalStateException("Serde class not found: ", e);
+			}
+			outboundBindTarget.to((Serde<Object>) keySerde, (Serde<Object>) valueSerde, name);
```

Comment:
![image](https://user-images.githubusercontent.com/1071590/28688276-563c8110-72df-11e7-9cca-cd4040adc61c.png)


That's might not be critical, but I guess some warn in logs during build is present.